web/src/main/java/org/springframework/security/web/reactive/result/method/annotation/CurrentSecurityContextArgumentResolver.java

```diff
@@ -0,0 +1,184 @@
+		if (reactiveSecurityContext == null) {
+			return null;
+		}
+		return reactiveSecurityContext.flatMap( a -> {
+			Object p = resolveSecurityContext(parameter, a);
+			Mono<Object> o = Mono.justOrEmpty(p);
+			return adapter == null ? o : Mono.just(adapter.fromPublisher(o));
+		});
+
+	}
+
+	/**
+	 * resolve the expression from {@link CurrentSecurityContext} annotation to get the value.
+	 * @param parameter the method parameter.
+	 * @param securityContext the security context.
+	 * @return the resolved object from expression.
+	 */
+	private Object resolveSecurityContext(MethodParameter parameter, SecurityContext securityContext) {
+		CurrentSecurityContext securityContextAnnotation = findMethodAnnotation(
+				CurrentSecurityContext.class, parameter);
+
+		Object securityContextResult = securityContext;
+
+		String expressionToParse = securityContextAnnotation.expression();
+		if (StringUtils.hasLength(expressionToParse)) {
+			StandardEvaluationContext context = new StandardEvaluationContext();
+			context.setRootObject(securityContext);
+			context.setVariable("this", securityContext);
+			context.setBeanResolver(beanResolver);
+
+			Expression expression = this.parser.parseExpression(expressionToParse);
+			securityContextResult = expression.getValue(context);
+		}
+
+		if (isInvalidType(parameter, securityContextResult)) {
+			if (securityContextAnnotation.errorOnInvalidType()) {
+				throw new ClassCastException(securityContextResult + " is not assignable to "
+						+ parameter.getParameterType());
+			}
+			else {
+				return null;
+			}
+		}
+
+		return securityContextResult;
+	}
+
+	/**
+	 * check if the retrieved value match with the parameter type.
+	 * @param parameter the method parameter.
+	 * @param reactiveSecurityContext the security context.
+	 * @return true = is not invalid type.
+	 */
+	private boolean isInvalidType(MethodParameter parameter, Object reactiveSecurityContext) {
+		if (reactiveSecurityContext == null) {
+			return false;
+		}
+		Class<?> typeToCheck = parameter.getParameterType();
+		boolean isParameterPublisher = Publisher.class.isAssignableFrom(parameter.getParameterType());
+		if (isParameterPublisher) {
+			ResolvableType resolvableType = ResolvableType.forMethodParameter(parameter);
+			Class<?> genericType = resolvableType.resolveGeneric(0);
+			if (genericType == null) {
+				return false;
+			}
+			typeToCheck = genericType;
+		}
+		return !typeToCheck.isAssignableFrom(reactiveSecurityContext.getClass());
+	}
+
+	/**
+	 * Obtains the specified {@link Annotation} on the specified {@link MethodParameter}.
+	 *
+	 * @param annotationClass the class of the {@link Annotation} to find on the
+	 * {@link MethodParameter}
+	 * @param parameter the {@link MethodParameter} to search for an {@link Annotation}
+	 * @return the {@link Annotation} that was found or null.
+	 */
+	private <T extends Annotation> T findMethodAnnotation(Class<T> annotationClass,
```

Comment:
@rwinch, it may be a good time to consider whether this ought to be moved elsewhere since it's in four places now. What do you think?

Reply:
yes, i think it is good to move to `AnnotationUtils`, :blush: